flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/read/HiveContinuousMonitoringFunction.java

```diff
@@ -0,0 +1,331 @@
+
+			@Override
+			public List<String> partitionKeys() {
+				return partitionKeys;
+			}
+
+			@Override
+			public Partition getPartition(String escapeName) throws TException {
+				return client.getPartition(
+						tablePath.getDatabaseName(),
+						tablePath.getObjectName(),
+						escapeName);
+			}
+
+			@Override
+			public List<Partition> listPartitionsByFilter(String filter) throws TException {
+				return client.listPartitionsByFilter(
+						tablePath.getDatabaseName(),
+						tablePath.getObjectName(),
+						filter,
+						(short) -1);
+			}
+
+			@Override
+			public FileSystem fileSystem() {
+				return fs;
+			}
+
+			@Override
+			public Path tableLocation() {
+				return new Path(hiveTable.getSd().getLocation());
+			}
+		};
+
+		this.distinctPartitions = new HashSet<>();
+		if (context.isRestored()) {
+			LOG.info("Restoring state for the {}.", getClass().getSimpleName());
+			this.currentReadTime = this.currReadTimeState.get().iterator().next();
+			this.distinctPartitions.addAll(this.distinctPartsState.get().iterator().next());
+		} else {
+			LOG.info("No state to restore for the {}.", getClass().getSimpleName());
+			if (startupPartition != null) {
+				this.currentReadTime = fetcher.extractTimestamp(
+						fetcherContext,
+						client.getPartition(
+								tablePath.getDatabaseName(),
+								tablePath.getObjectName(),
+								escapePathName(startupPartition)));
+			}
+		}
+	}
+
+	@Override
+	public void run(SourceContext<TimestampedHiveInputSplit> context) throws Exception {
+		checkpointLock = context.getCheckpointLock();
+		while (isRunning) {
+			synchronized (checkpointLock) {
+				monitorAndForwardSplits(context);
+			}
+			Thread.sleep(interval);
+		}
+	}
+
+	private void monitorAndForwardSplits(
+			SourceContext<TimestampedHiveInputSplit> context) throws Exception {
+		assert (Thread.holdsLock(checkpointLock));
+
+		List<Partition> partitions = fetcher.fetchPartitions(fetcherContext, currentReadTime);
+
+		if (partitions.isEmpty()) {
+			return;
+		}
+
+		long maxTimestamp = Long.MIN_VALUE;
+		Set<List<String>> nextDistinctParts = new HashSet<>();
+		for (Partition partition : partitions) {
+			List<String> partSpec = partition.getValues();
+			if (!this.distinctPartitions.contains(partSpec)) {
+				this.distinctPartitions.add(partition.getValues());
+				long timestamp = this.fetcher.extractTimestamp(fetcherContext, partition);
+				if (timestamp > currentReadTime) {
+					nextDistinctParts.add(partSpec);
+				}
+				if (timestamp > maxTimestamp) {
+					maxTimestamp = timestamp;
+				}
+				HiveTableInputSplit[] splits = HiveTableInputFormat.createInputSplits(
+						this.readerParallelism,
+						Collections.singletonList(toHiveTablePartition(partition)),
+						this.conf.conf());
+				for (HiveTableInputSplit split : splits) {
+					context.collect(new TimestampedHiveInputSplit(timestamp, split));
+				}
+			}
+		}
+
+		if (maxTimestamp > currentReadTime) {
```

Comment:
Just ask the similar question as @lirui-apache said,   if the partitions is not sorted by timestamp ascending, then it seems have problems,  say we have partitions with timestamp like `[(p0, 5), (p1, 1), (p2, 3)]` ( the first part of (p0, 5) is partition name and the second is timestamp ),   now we have emitted the timestamp=5 successfully, then we will update the currentReadTime = 5   I think. Once the operator restored, we will just start from the timestamp = 5 to read the following partitions...it mistakenly skipped other timestamps such as 1 and 3 , finally we will lost part of the data to consume....
Pls correct me if I'm wrong.

Reply:
First, will sort this list.

Reply:
> we will lost part of the data to consume

No, these partitions will be all emitted to downstream, there is no checkpoint in them.